db/design_doc.go

```diff
@@ -265,3 +287,396 @@ func stripSyncProperty(row *sgbucket.ViewRow) {
+		                            emit(name, value)
+		                        }
+
+		                    }
+		               }`
+	roleAccess_vbSeq_map = fmt.Sprintf(roleAccess_vbSeq_map, syncData)
+
+	designDocMap := map[string]sgbucket.DesignDoc{}
+	designDocMap[DesignDocSyncGateway()] = sgbucket.DesignDoc{
+		Views: sgbucket.ViewMap{
+			ViewChannels:        sgbucket.ViewDef{Map: channels_map},
+			ViewAccess:          sgbucket.ViewDef{Map: access_map},
+			ViewRoleAccess:      sgbucket.ViewDef{Map: roleAccess_map},
+			ViewAccessVbSeq:     sgbucket.ViewDef{Map: access_vbSeq_map},
+			ViewRoleAccessVbSeq: sgbucket.ViewDef{Map: roleAccess_vbSeq_map},
+			ViewPrincipals:      sgbucket.ViewDef{Map: principals_map},
+		},
+		Options: &sgbucket.DesignDocOptions{
+			IndexXattrOnTombstones: true,
+		},
+	}
+
+	designDocMap[DesignDocSyncHousekeeping()] = sgbucket.DesignDoc{
+		Views: sgbucket.ViewMap{
+			ViewAllBits:    sgbucket.ViewDef{Map: allbits_map},
+			ViewAllDocs:    sgbucket.ViewDef{Map: alldocs_map, Reduce: "_count"},
+			ViewImport:     sgbucket.ViewDef{Map: import_map, Reduce: "_count"},
+			ViewOldRevs:    sgbucket.ViewDef{Map: oldrevs_map, Reduce: "_count"},
+			ViewSessions:   sgbucket.ViewDef{Map: sessions_map},
+			ViewTombstones: sgbucket.ViewDef{Map: tombstones_map},
+		},
+		Options: &sgbucket.DesignDocOptions{
+			IndexXattrOnTombstones: true, // For ViewTombstones
+		},
+	}
+
+	sleeper := base.CreateDoublingSleeperFunc(
+		11, //MaxNumRetries approx 10 seconds total retry duration
+		5,  //InitialRetrySleepTimeMS
+	)
+
+	// add all design docs from map into bucket
+	for designDocName, designDoc := range designDocMap {
+
+		//start a retry loop to put design document backing off double the delay each time
+		worker := func() (shouldRetry bool, err error, value interface{}) {
+			err = bucket.PutDDoc(designDocName, designDoc)
+			if err != nil {
+				base.Warn("Error installing Couchbase design doc: %v", err)
+			}
+			return err != nil, err, nil
+		}
+
+		description := fmt.Sprintf("Attempt to install Couchbase design doc bucket : %v", designDocName)
+		err, _ := base.RetryLoop(description, worker, sleeper)
+
+		if err != nil {
+			return pkgerrors.Wrapf(err, "Error installing Couchbase Design doc: %v", designDocName)
+		}
+	}
+
+	base.Logf("Design docs successfully created for view version %s.", DesignDocVersion)
+
+	return nil
+}
+
+// Issue a stale=false queries against critical views to guarantee indexing is complete and views are ready
+func WaitForViews(bucket base.Bucket) error {
+	viewErrors := make(chan error, 3)
+	var viewsWg sync.WaitGroup
+	views := []string{ViewChannels, ViewAccess, ViewRoleAccess}
+
+	base.Logf("Verifying view availability for bucket %s...", bucket.GetName())
+
+	for _, viewName := range views {
+		viewsWg.Add(1)
+		go func(view string) {
+			defer viewsWg.Done()
+			viewErr := waitForViewIndexing(bucket, DesignDocSyncGateway(), view)
+			if viewErr != nil {
+				viewErrors <- viewErr
```

Comment:
Would it help support/debugging if this error were to include the name of the view that failed? (Either here, or further down the `waitForViewIndexing` stack)